content/blog/blog-post-09.md

```diff
@@ -0,0 +1,37 @@
+---
+authors: ["kokkos-team"]
+title: "Kokkos Releases New Autotuning Features"
+date: 2024-12-20
+tags: ["blog"]
+thumbnail: img/blog/apex-kokkos.png
+---
+
+# Motivation
+
+By default, internal Kokkos execution space parameters are empirically or heuristically hand-tuned with fixed parameter values to provide "one size fits most" performance, with the goal of minimizing the effect of the abstraction overhead and approximating the performance of an optimized, lower-level implementation. Can these parameters be tuned for a particular application and architecture for programmers to easily tackle further performance opportunities? The Kokkos Tools APEX auto-tuning connector [5], a git submodule in Kokkos Tools with a stable version released in Kokkos 4.5 [4], offers an answer.
+
+# How it Works
+
+Kokkos includes a Tuning API (TuningInterface) that can be used to construct a tuning context around a computational kernel, declare input variables that define the context state, declare output variables to be tuned, and request output variables when the kernel is executed. The Kokkos Tools infrastructure provide integrated support to utilize this API during Kokkos application execution, i.e., online, rather than offline [2]. Together, Kokkos Tools and the Tuning API is used in APEX to tune at runtime Kokkos kernel parameters running in any execution space / policy combination. We note through this Kokkos auto-tuning capability from APEX allows for (a) switching its tuning heuristics between Kokkos Execution Spaces (i.e. choose between serial or OpenMP depending on the problem size, etc.) or execution policies and (b) auto-tuning any arbitrary parameter within an application that uses Kokkos - solver choices, algorithmic parameters, tolerances, etc.
```

Comment:
```suggestion
Kokkos includes a Tuning API (TuningInterface) that can be used to construct a tuning context around a computational kernel, declare input variables that define the context state, declare output variables to be tuned, and request output variables when the kernel is executed. The Kokkos Tools infrastructure provide integrated support to utilize this API during Kokkos application execution, i.e., online, rather than offline [2]. Together, Kokkos Tools and the Tuning API is used in APEX to tune at runtime Kokkos kernel parameters running in any execution space / policy combination. We note through this Kokkos auto-tuning capability from APEX allows for (a) switching its tuning heuristics between Kokkos Execution Spaces (i.e. choose between Serial or OpenMP depending on the problem size, etc.) or execution policies and (b) auto-tuning any arbitrary parameter within an application that uses Kokkos - solver choices, algorithmic parameters, tolerances, etc.
```

Reply:
I wonder if we can come up with a code snippet illustrating how one declare tunable variables and whether it would help here.  Thoughts?

Reply:
> I wonder if we can come up with a code snippet illustrating how one declare tunable variables and whether it would help here. Thoughts?

Yes, I think we could show the matrix multiplication code that is being experimented with. The code and setup with the Tuning API is a bit long (about 50 lines of code) and all of it is needed. @khuck and I discussed it but decided it's not needed to be shown here - maybe we can go into more detail in that Wiki Post we linked in the Outcomes Section.

Reply:
> What is "TuningInterface" referring to?


This is what @khuck and I used to refer to the set of functions in Kokkos::Tools that helps declare tuning parameter variables and tuning contexts in which those variables should be tuned, e.g., the function Kokkos::Tools::declareTuningVariable() is part of that interface. 

I think the text TuningInterface could actually be taken out and we can link to the Wiki Post for more information. 

Also, maybe this can documented better in Kokkos or Kokkos Tools right now - this GitHub issue is probably the best documentation for its use: https://github.com/kokkos/kokkos-tools/issues/90

Reply:
> I wonder if we can come up with a code snippet illustrating how one declare tunable variables and whether it would help here. Thoughts?

Well... that's one of the issues with the current API, it is kind of complicated to create a variable...